File: app/src/main/java/org/mifos/mobilebanking/presenters/QrCodeImportPresenter.java
package org.mifos.mobilebanking.presenters;

import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;


import com.google.zxing.BinaryBitmap;
import com.google.zxing.DecodeHintType;
import com.google.zxing.LuminanceSource;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.RGBLuminanceSource;
import com.google.zxing.Reader;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import com.isseiaoki.simplecropview.CropImageView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.injection.ApplicationContext;
import org.mifos.mobilebanking.presenters.base.BasePresenter;
import org.mifos.mobilebanking.ui.views.QrCodeImportView;

import java.util.EnumMap;
import java.util.Map;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by manishkumar on 19/05/18.
 */

public class QrCodeImportPresenter extends BasePresenter<QrCodeImportView> {

    private Result result;
    private CompositeDisposable compositeDisposable;

    /**
     * Initialises the LoanAccountDetailsPresenter by automatically injecting an instance of
     * {@link Context}.
     *
     * @param context Context of the view attached to the presenter. In this case
     *                it is that of an {@link android.support.v7.app.AppCompatActivity}
     */
    @Inject
    public QrCodeImportPresenter(@ApplicationContext Context context) {
        super(context);
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(QrCodeImportView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
        context = null;
    }

    /**
     * This method decodes the Qr code from the selected region
     *
     * @param sourceUri     contains the Uri of qr code from gallery
     * @param cropImageView contains cropImageView of layout
     */
    public void getDecodedResult(Uri sourceUri, final CropImageView cropImageView) {
        checkViewAttached();
        getMvpView().showProgress();
        compositeDisposable.add(cropImageView.crop(sourceUri)
                .executeAsSingle()
                .flatMap(new Function<Bitmap, SingleSource<Result>>() {
                    @Override
                    public SingleSource<Result> apply(@NonNull Bitmap bMap)
                            throws Exception {

                        int[] intArray = new int[bMap.getWidth() * bMap.getHeight()];
                        //copy pixel data from the Bitmap into the 'intArray' array
                        bMap.getPixels(intArray, 0, bMap.getWidth(), 0, 0,
                                bMap.getWidth(), bMap.getHeight());
                        LuminanceSource source = new RGBLuminanceSource(bMap.getWidth(),
                                bMap.getHeight(), intArray);
                        BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));
                        //flags for decoding since it is large file
                        Map<DecodeHintType, Boolean> tmpHintsMap
                                = new EnumMap<>(DecodeHintType.class);
                        tmpHintsMap.put(DecodeHintType.TRY_HARDER, Boolean.TRUE);
                        tmpHintsMap.put(DecodeHintType.PURE_BARCODE, Boolean.TRUE);

                        Reader reader = new MultiFormatReader();

                        try {
                            result = reader.decode(bitmap, tmpHintsMap);
                        } catch (Exception e) {
                            getMvpView().hideProgress();
                            getMvpView().showErrorReadingQr(context
                                    .getString(R.string.error_reading_qr));
                        }
                        return Single.just(result);
                    }
                })
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<Result>() {
                    @Override
                    public void accept(Result result) throws Exception {
                        getMvpView().hideProgress();
                        getMvpView().handleDecodedResult(result);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        getMvpView().hideProgress();
                        getMvpView()
                                .showErrorReadingQr(context.getString(R.string.error_reading_qr));
                    }
                }));

    }


}

File: app/src/main/java/org/mifos/mobile/presenters/QrCodeImportPresenter.kt
package org.mifos.mobile.presenters

import android.content.Context
import android.net.Uri

import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer

import com.isseiaoki.simplecropview.CropImageView

import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

import org.mifos.mobile.R
import org.mifos.mobile.injection.ApplicationContext
import org.mifos.mobile.presenters.base.BasePresenter
import org.mifos.mobile.ui.views.QrCodeImportView

import java.util.*
import javax.inject.Inject

/**
 * Created by manishkumar on 19/05/18.
 */
class QrCodeImportPresenter @Inject constructor(@ApplicationContext context: Context?) :
        BasePresenter<QrCodeImportView?>(context) {

    private var result: Result? = null
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    private var hasErrorOccured = false

    override fun attachView(mvpView: QrCodeImportView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        compositeDisposable.clear()
        context = null
    }

    /**
     * This method decodes the Qr code from the selected region
     *
     * @param sourceUri     contains the Uri of qr code from gallery
     * @param cropImageView contains cropImageView of layout
     */
    fun getDecodedResult(sourceUri: Uri?, cropImageView: CropImageView?) {
        checkViewAttached()
        mvpView?.showProgress()
        cropImageView?.crop(sourceUri)
                ?.executeAsSingle()
                ?.flatMap { bMap ->
                    val intArray = IntArray(bMap.width * bMap.height)
                    //copy pixel data from the Bitmap into the 'intArray' array
                    bMap.getPixels(intArray, 0, bMap.width, 0, 0,
                            bMap.width, bMap.height)
                    val source: LuminanceSource = RGBLuminanceSource(bMap.width,
                            bMap.height, intArray)
                    val bitmap = BinaryBitmap(HybridBinarizer(source))
                    //flags for decoding since it is large file
                    val tmpHintsMap: MutableMap<DecodeHintType, Boolean?> = EnumMap(DecodeHintType::class.java)
                    tmpHintsMap[DecodeHintType.TRY_HARDER] = java.lang.Boolean.TRUE
                    tmpHintsMap[DecodeHintType.PURE_BARCODE] = java.lang.Boolean.TRUE
                    val reader: Reader = MultiFormatReader()
                    try {
                        result = reader.decode(bitmap, tmpHintsMap)
                    } catch (e: Exception) {
                        mvpView?.hideProgress()
                        hasErrorOccured = true
                        mvpView?.showErrorReadingQr(context
                                ?.getString(R.string.error_reading_qr))
                    }
                    Single.just(result)
                }
                ?.subscribeOn(Schedulers.newThread())
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribe({ result ->
                    mvpView?.hideProgress()
                    mvpView?.handleDecodedResult(result)
                }) {
                    mvpView?.hideProgress()
                    if (!hasErrorOccured) {
                        mvpView
                                ?.showErrorReadingQr(
                                        context?.getString(R.string.error_reading_qr))
                    }
                    hasErrorOccured = false
                }?.let { compositeDisposable.add(it) }
    }

}